jerry-main/fuzzilli.c

```diff
@@ -0,0 +1,371 @@
+
+      jerry_release_value(str_val);
+    }
+
+    if (args_cnt == 0 || jerry_value_is_error(ret_val))
+    {
+      putc('\n', fzliout);
+    }
+    fflush(fzliout);
+  }
+
+  jerry_heap_free(operation, str_size * sizeof(jerry_char_t));
+  return ret_val;
+}
+
+/**
+ * Register a JavaScript function in the global object.
+ */
+static void
+register_js_function(const char *name_p,                 /**< name of the function */
+                     jerry_external_handler_t handler_p) /**< function callback */
+{
+  jerry_value_t result_val = jerryx_handler_register_global((const jerry_char_t *)name_p, handler_p);
+
+  if (jerry_value_is_error(result_val))
+  {
+    jerry_port_log(JERRY_LOG_LEVEL_WARNING, "Warning: failed to register '%s' method.", name_p);
+    result_val = jerry_get_value_from_error(result_val, true);
+  }
+
+  jerry_release_value(result_val);
+} /* register_js_function */
+
+int main ()
+{
+  // Let parent know we are ready
+  char helo[] = "HELO";
+
+  if (write(REPRL_CWFD, helo, 4) != 4 ||
+      read(REPRL_CRFD, helo, 4) != 4 ||
+      memcmp(helo, "HELO", 4) != 0)
+  {
+    fprintf(stderr, "Invalid response from parent\n");
+    _exit(-1);
+  }
+
+  while (true) {
+    jerry_init(JERRY_INIT_EMPTY);
+    // Register a JavaScript function 'fuzzilli' in the global object.
+    register_js_function("fuzzilli", jerryx_handler_fuzzilli);
+
+    unsigned action = 0;
+    ssize_t nread = read(REPRL_CRFD, &action, 4);
+    fflush(0);
+    if (nread != 4 || action != 0x63657865) // exec
+    {
+      fprintf(stderr, "Unknown action: %x\n", action);
+      _exit(-1);
+    }
+
+    size_t script_size = 0;
+    uint8_t *source_buffer_tail = buffer;
+    size_t len = 0;
+
+    read(REPRL_CRFD, &script_size, 8);
+    source_buffer_tail = buffer;
+    ssize_t remaining = (ssize_t)script_size;
+    while (remaining > 0)
+    {
+      ssize_t rv = read(REPRL_DRFD, source_buffer_tail, (size_t)remaining);
+      if (rv <= 0)
+      {
+        fprintf(stderr, "Failed to load script\n");
+        _exit(-1);
+      }
+      remaining -= rv;
+      source_buffer_tail += rv;
+    }
+    buffer[script_size] = 0;
+    len = script_size;
+
+    if (len > 0 && jerry_is_valid_utf8_string (buffer, (jerry_size_t)len))
+    {
+      int status_rc = 0;
+
+      jerry_value_t parse_value = jerry_parse(NULL, 0, (jerry_char_t *)buffer, len, JERRY_PARSE_NO_OPTS);
+      jerry_value_t run_value;
+
+      if (!jerry_value_is_error(parse_value))
+      {
+        run_value = jerry_run(parse_value);
+
+        if (!jerry_value_is_error (run_value))
+        {
+          jerry_value_t run_queue_value = jerry_run_all_enqueued_jobs ();
+          jerry_release_value(run_queue_value);
+        }
+        else
+        {
+          status_rc = 1;
+        }
+        jerry_release_value(run_value);
+      }
+
+      jerry_release_value(parse_value);
+      jerry_cleanup();
+
+      status_rc <<= 8;
+      if (write(REPRL_CWFD, &status_rc, 4) != 4)
```

Comment:
This is "only" a style issue, but my eyes miss the curly braces. (Doesn't our style checker catch this?...)

Reply:
No. As I am not acquainted with the building process (nor codebase), I found `check-vera.sh` only after some tests failed, so I had to correct about 100+ another styling issues, it should be fine now.

Anyway, thanks for your time and valuable feedback! I fixed most of the things I was able, but if you think this requires too much work to maintain (or adjust to conform the JerryScript convention), I am fine with keeping the integration part as an external patch.

Reply:
>  but if you think this requires too much work to maintain (or adjust to conform the JerryScript convention), I am fine with keeping the integration part as an external patch.

Nah, that was not my intent with the review. External patches bit rot. Upstream maintainers don't know about them so don't care about them. Patch maintainers often don't have the time to follow up on upstream. It's better to have this here, JerryScript also benefits from this.

We are getting there, I think we may have the right approach with the third-party subdir. Just bear with me, I'm admittedly nitpicky.